internal/querycoordv2/job/load_job.go

```diff
@@ -0,0 +1,389 @@
+		BaseJob:        NewBaseJob(ctx, req.Base.GetMsgID(), req.GetCollectionID()),
+		req:            req,
+		undo:           NewUndoList(ctx, meta, cluster, targetMgr, targetObserver),
+		dist:           dist,
+		meta:           meta,
+		cluster:        cluster,
+		targetMgr:      targetMgr,
+		targetObserver: targetObserver,
+		broker:         broker,
+		nodeMgr:        nodeMgr,
+	}
+}
+
+func (job *LoadCollectionJob) PreExecute() error {
+	req := job.req
+	log := log.Ctx(job.ctx).With(zap.Int64("collectionID", req.GetCollectionID()))
+
+	if req.GetReplicaNumber() <= 0 {
+		log.Info("request doesn't indicate the number of replicas, set it to 1",
+			zap.Int32("replicaNumber", req.GetReplicaNumber()))
+		req.ReplicaNumber = 1
+	}
+
+	collection := job.meta.GetCollection(req.GetCollectionID())
+	if collection == nil {
+		return nil
+	}
+
+	if collection.GetReplicaNumber() != req.GetReplicaNumber() {
+		msg := fmt.Sprintf("collection with different replica number %d existed, release this collection first before changing its replica number",
+			job.meta.GetReplicaNumber(req.GetCollectionID()),
+		)
+		log.Warn(msg)
+		return utils.WrapError(msg, ErrLoadParameterMismatched)
+	} else if !typeutil.MapEqual(collection.GetFieldIndexID(), req.GetFieldIndexID()) {
+		msg := fmt.Sprintf("collection with different index %v existed, release this collection first before changing its index",
+			collection.GetFieldIndexID())
+		log.Warn(msg)
+		return utils.WrapError(msg, ErrLoadParameterMismatched)
+	}
+
+	return nil
+}
+
+func (job *LoadCollectionJob) Execute() error {
+	req := job.req
+	log := log.Ctx(job.ctx).With(zap.Int64("collectionID", req.GetCollectionID()))
+	meta.GlobalFailedLoadCache.Remove(req.GetCollectionID())
+
+	// Fetch partitions from RootCoord
+	partitionIDs, err := job.broker.GetPartitions(job.ctx, req.GetCollectionID())
+	if err != nil {
+		msg := "failed to get partitions from RootCoord"
+		log.Error(msg, zap.Error(err))
+		return utils.WrapError(msg, err)
+	}
+	loadedPartitionIDs := lo.Map(job.meta.CollectionManager.GetPartitionsByCollection(req.GetCollectionID()),
+		func(partition *meta.Partition, _ int) int64 {
+			return partition.GetPartitionID()
+		})
+	lackPartitionIDs := lo.FilterMap(partitionIDs, func(partID int64, _ int) (int64, bool) {
+		return partID, !lo.Contains(loadedPartitionIDs, partID)
+	})
+	if len(lackPartitionIDs) == 0 {
+		return ErrCollectionLoaded
+	}
+	job.undo.CollectionID = req.GetCollectionID()
+	job.undo.LackPartitions = lackPartitionIDs
+	log.Info("find partitions to load", zap.Int64s("partitions", lackPartitionIDs))
+
+	_, err = job.targetObserver.UpdateNextTarget(req.GetCollectionID(), append(loadedPartitionIDs, lackPartitionIDs...)...)
+	if err != nil {
+		msg := "failed to update next target"
+		log.Error(msg, zap.Error(err))
+		return utils.WrapError(msg, err)
+	}
+	job.undo.TargetUpdated = true
+
+	err = loadPartitions(job.ctx, job.meta, job.cluster, false, req.GetCollectionID(), lackPartitionIDs...)
```

Comment:
Also I guess this will not work if it's the first time to load the collection

Reply:
> Also I guess this will not work if it's the first time to load the collection

yep, I think it's ok to not work for the first time, as these partitions would be created during the watchDM task in QueryNodes.

Reply:
> Notifing QueryNodes before updating target would be better, so the QueryNodes will never loss the data of new partitions, not depends on the impl

QueryNodes will receive all data from the streaming regardless of whether or not the partitions have been loaded.